File: tool/convert_to_latex.dart
import 'dart:convert';
import 'dart:io';

/// Pass 1: Replace Unicode characters with LaTeX commands (no $ wrapping)
String unicodeToLatex(String text) {
  if (text.isEmpty) return text;
  
  final buf = StringBuffer();
  final runes = text.runes.toList();
  int i = 0;
  
  while (i < runes.length) {
    final ch = String.fromCharCode(runes[i]);
    
    // Superscripts - collect consecutive
    if (_superscripts.containsKey(ch)) {
      String sup = '';
      while (i < runes.length && _superscripts.containsKey(String.fromCharCode(runes[i]))) {
        sup += _superscripts[String.fromCharCode(runes[i])]!;
        i++;
      }
      buf.write(sup.length > 1 ? '^{$sup}' : '^$sup');
      continue;
    }
    
    // Subscripts - collect consecutive
    if (_subscripts.containsKey(ch)) {
      String sub = '';
      while (i < runes.length && _subscripts.containsKey(String.fromCharCode(runes[i]))) {
        sub += _subscripts[String.fromCharCode(runes[i])]!;
        i++;
      }
      buf.write(sub.length > 1 ? '_{$sub}' : '_$sub');
      continue;
    }
    
    // Check replacement maps
    if (_replacements.containsKey(ch)) {
      buf.write(_replacements[ch]);
      i++;
      continue;
    }
    
    buf.write(ch);
    i++;
  }
  
  return buf.toString();
}

/// Pass 2: Wrap math expressions in $...$
/// Strategy: tokenize by spaces, classify each token, merge adjacent math tokens
String wrapMathRegions(String text) {
  if (text.isEmpty) return text;
  
  // Don't process if already has $ (already LaTeX formatted)
  if (text.contains('\$')) return text;
  
  // Split into tokens preserving separators
  final tokens = <_Token>[];
  final pattern = RegExp(r'(\s+|[,;:!?]+(?:\s|$))');
  int pos = 0;
  
  for (final match in pattern.allMatches(text)) {
    if (match.start > pos) {
      tokens.add(_Token(text.substring(pos, match.start), false));
    }
    tokens.add(_Token(match.group(0)!, false)); // separator
    pos = match.end;
  }
  if (pos < text.length) {
    tokens.add(_Token(text.substring(pos), false));
  }
  
  // Classify each non-separator token as math or text
  for (int i = 0; i < tokens.length; i++) {
    final t = tokens[i];
    if (t.text.trim().isEmpty || RegExp(r'^[,;:!?]+\s*$').hasMatch(t.text)) continue;
    tokens[i] = _Token(t.text, _isMathToken(t.text));
  }
  
  // Build output: group consecutive math tokens and wrap in $...$
  final buf = StringBuffer();
  int ti = 0;
  
  while (ti < tokens.length) {
    if (!tokens[ti].isMath) {
      buf.write(tokens[ti].text);
      ti++;
      continue;
    }
    
    // Start of math region
    final mathParts = <String>[];
    mathParts.add(tokens[ti].text);
    ti++;
    
    // Continue collecting math and connecting separators
    while (ti < tokens.length) {
      // Check if next non-separator is also math
      if (tokens[ti].text.trim().isEmpty || RegExp(r'^[,;:!?]+\s*$').hasMatch(tokens[ti].text)) {
        // It's a separator - check if next content is math
        int nextContent = ti + 1;
        while (nextContent < tokens.length && 
               (tokens[nextContent].text.trim().isEmpty || 
                RegExp(r'^[,;:!?]+\s*$').hasMatch(tokens[nextContent].text))) {
          nextContent++;
        }
        
        if (nextContent < tokens.length && tokens[nextContent].isMath) {
          // Include separator in math region
          for (int j = ti; j <= nextContent; j++) {
            mathParts.add(tokens[j].text);
          }
          ti = nextContent + 1;
          continue;
        } else {
          break; // End of math region
        }
      } else if (tokens[ti].isMath) {
        mathParts.add(tokens[ti].text);
        ti++;
      } else {
        // Check if it's a short connecting word between math
        final word = tokens[ti].text.toLowerCase();
        if (_mathConnectors.contains(word) && ti + 1 < tokens.length) {
          // Look ahead for more math
          int nextContent = ti + 1;
          while (nextContent < tokens.length && 
                 (tokens[nextContent].text.trim().isEmpty || 
                  RegExp(r'^[,;:!?]+\s*$').hasMatch(tokens[nextContent].text))) {
            nextContent++;
          }
          if (nextContent < tokens.length && tokens[nextContent].isMath) {
            // Include connector in math region
            for (int j = ti; j <= nextContent; j++) {
              mathParts.add(tokens[j].text);
            }
            ti = nextContent + 1;
            continue;
          }
        }
        break; // End of math region
      }
    }
    
    final mathContent = mathParts.join('').trim();
    if (mathContent.isNotEmpty) {
      buf.write('\$$mathContent\$');
    }
  }
  
  String result = buf.toString();
  
  // Post-processing: fix common issues
  // Remove $ around very short non-math content
  result = result.replaceAllMapped(RegExp(r'\$([a-zA-Z])\$'), (m) {
    // Single letter that's probably a variable - keep it in math
    return '\$${m.group(1)}\$';
  });
  
  // Merge adjacent $...$ blocks with just whitespace between
  result = result.replaceAllMapped(RegExp(r'\$\s*\$'), (m) => ' ');
  
  // Fix "$x$ = $y$" to "$x = y$" 
  result = _mergeAdjacentMath(result);
  
  return result;
}

/// Merge adjacent math regions separated by operators or short connectors
String _mergeAdjacentMath(String text) {
  // Pattern: $...$ followed by optional space and operator/= and space and $...$
  final pattern = RegExp(r'\$([^$]+)\$(\s*[=<>+\-*/|&]\s*)\$([^$]+)\$');
  String prev = text;
  String current = text.replaceAllMapped(pattern, (m) => '\$${m.group(1)}${m.group(2)}${m.group(3)}\$');
  while (current != prev) {
    prev = current;
    current = current.replaceAllMapped(pattern, (m) => '\$${m.group(1)}${m.group(2)}${m.group(3)}\$');
  }
  return current;
}

/// Determine if a token is mathematical
bool _isMathToken(String token) {
  // Contains LaTeX commands
  if (token.contains('\\')) return true;
  
  // Contains ^ or _ (sub/superscript)
  if (token.contains('^') || token.contains('_')) return true;
  
  // Is a pure number
  if (RegExp(r'^-?\d+(/\d+)?$').hasMatch(token)) return false; // plain numbers in text
  
  // Contains mixed letters and numbers with math operators
  if (RegExp(r'[A-Za-z]\d|[=<>]').hasMatch(token) && 
      RegExp(r'[\^_\\=<>+\-*/]').hasMatch(token)) return true;
  
  // Is a single variable letter followed by sub/super
  if (RegExp(r'^[A-Za-z][\^_]').hasMatch(token)) return true;
  
  // Contains mathematical operators after letters
  if (RegExp(r'[A-Za-z][=<>+\-*/][A-Za-z0-9]').hasMatch(token)) return true;
  
  // Common math patterns
  if (RegExp(r'^\(?[A-Za-z]\)?[=<>]').hasMatch(token)) return true;
  
  // Fractions like 1/n, x/y
  if (RegExp(r'^[A-Za-z0-9]+/[A-Za-z0-9]+$').hasMatch(token) && 
      RegExp(r'[A-Za-z]').hasMatch(token)) return true;
  
  // Parenthesized expressions with variables: (1-p), (n+1), etc.
  if (RegExp(r'^\(.*[A-Za-z].*[+\-*/].*\)$').hasMatch(token)) return true;
  
  // Matrix notation [[...]]
  if (token.contains('[[')) return true;
  
  return false;
}

const _mathConnectors = {'et', 'ou', 'si', 'car', 'avec', 'pour', 'and', 'or', 'where', 'mod'};

class _Token {
  final String text;
  final bool isMath;
  _Token(this.text, this.isMath);
}

/// Combined conversion
String convertToLatex(String text) {
  if (text.isEmpty) return text;
  
  // Skip if already contains LaTeX math delimiters
  if (text.contains('\$')) return text;
  
  // Pass 1: Unicode to LaTeX commands
  String result = unicodeToLatex(text);
  
  // Pass 2: Wrap math in $...$
  result = wrapMathRegions(result);
  
  return result;
}

// =================== Character maps ===================

const _superscripts = {
  '⁰': '0', '¹': '1', '²': '2', '³': '3', '⁴': '4',
  '⁵': '5', '⁶': '6', '⁷': '7', '⁸': '8', '⁹': '9',
  'ⁿ': 'n', 'ᵏ': 'k', 'ᵐ': 'm', 'ᵖ': 'p', 'ᵍ': 'q',
  '⁻': '-', '⁺': '+',
};

const _subscripts = {
  '₀': '0', '₁': '1', '₂': '2', '₃': '3', '₄': '4',
  '₅': '5', '₆': '6', '₇': '7', '₈': '8', '₉': '9',
  'ₙ': 'n', 'ₖ': 'k', 'ₘ': 'm', 'ₚ': 'p', 'ₐ': 'a',
  'ᵢ': 'i', 'ⱼ': 'j',
};

const _replacements = {
  'ℝ': r'\mathbb{R}',
  'ℤ': r'\mathbb{Z}',
  'ℚ': r'\mathbb{Q}',
  'ℂ': r'\mathbb{C}',
  'ℕ': r'\mathbb{N}',
  '𝔽': r'\mathbb{F}',
  '𝕋': r'\mathbb{T}',
  '𝟙': r'\mathbf{1}',
  '∈': r'\in ',
  '∉': r'\notin ',
  '≤': r'\leq ',
  '≥': r'\geq ',
  '≠': r'\neq ',
  '≡': r'\equiv ',
  '≅': r'\simeq ',
  '≈': r'\approx ',
  '⟹': r'\Rightarrow ',
  '⟸': r'\Leftarrow ',
  '⟺': r'\Leftrightarrow ',
  '⇒': r'\Rightarrow ',
  '⇐': r'\Leftarrow ',
  '⇔': r'\Leftrightarrow ',
  '→': r'\to ',
  '←': r'\leftarrow ',
  '↦': r'\mapsto ',
  '↗': r'\nearrow ',
  '∀': r'\forall ',
  '∃': r'\exists ',
  '∅': r'\emptyset',
  '∞': r'\infty',
  '∪': r'\cup ',
  '∩': r'\cap ',
  '⊂': r'\subset ',
  '⊃': r'\supset ',
  '⊆': r'\subseteq ',
  '⊇': r'\supseteq ',
  '⊕': r'\oplus ',
  '⊗': r'\otimes ',
  '⊥': r'\perp',
  '⟨': r'\langle ',
  '⟩': r'\rangle ',
  '‖': r'\|',
  '∫': r'\int',
  '∑': r'\sum',
  '∏': r'\prod',
  '√': r'\sqrt',
  '⌊': r'\lfloor ',
  '⌋': r'\rfloor ',
  '·': r'\cdot ',
  '×': r'\times ',
  '≢': r'\not\equiv ',
  'λ': r'\lambda',
  'μ': r'\mu',
  'σ': r'\sigma',
  'ε': r'\varepsilon',
  'δ': r'\delta',
  'φ': r'\varphi',
  'ω': r'\omega',
  'θ': r'\theta',
  'α': r'\alpha',
  'β': r'\beta',
  'γ': r'\gamma',
  'Δ': r'\Delta',
  'Φ': r'\Phi',
  'ξ': r'\xi',
  'ψ': r'\psi',
  'ρ': r'\rho',
  'τ': r'\tau',
  'ν': r'\nu',
  'χ': r'\chi',
  'ℓ': r'\ell',
  // Note: π kept as-is when used in text like "π/2" (will be in math context)
  'π': r'\pi',
};

// =================== File processors ===================

void processFile(String path, String key, Function(Map<String, dynamic>) processor) {
  print('Processing $path...');
  final file = File(path);
  final content = file.readAsStringSync();
  final data = jsonDecode(content) as Map<String, dynamic>;
  
  if (data.containsKey(key)) {
    final list = data[key] as List;
    for (int i = 0; i < list.length; i++) {
      processor(list[i] as Map<String, dynamic>);
    }
    print('  Processed ${list.length} entries.');
  }
  
  final encoder = JsonEncoder.withIndent('  ');
  file.writeAsStringSync(encoder.convert(data));
}

void processExercise(Map<String, dynamic> ex) {
  if (ex['enonce'] is String) ex['enonce'] = convertToLatex(ex['enonce']);
  if (ex['indication'] is String) ex['indication'] = convertToLatex(ex['indication']);
  if (ex['solution'] is List) {
    ex['solution'] = (ex['solution'] as List).map((s) => 
      s is String ? convertToLatex(s) : s).toList();
  }
}

void processDemo(Map<String, dynamic> demo) {
  if (demo['enonce'] is String) demo['enonce'] = convertToLatex(demo['enonce']);
  if (demo['preuve'] is List) {
    demo['preuve'] = (demo['preuve'] as List).map((s) => 
      s is String ? convertToLatex(s) : s).toList();
  }
  if (demo['corollaires'] is List) {
    demo['corollaires'] = (demo['corollaires'] as List).map((s) => 
      s is String ? convertToLatex(s) : s).toList();
  }
}

void processContreExemple(Map<String, dynamic> ce) {
  if (ce['enonce'] is String) ce['enonce'] = convertToLatex(ce['enonce']);
  if (ce['explication'] is String) ce['explication'] = convertToLatex(ce['explication']);
}

void main() {
  final baseDir = Directory.current.path;
  final dataDir = '$baseDir/assets/data';
  
  processFile('$dataDir/exercices.json', 'exercices', processExercise);
  processFile('$dataDir/demonstrations.json', 'demonstrations', processDemo);
  processFile('$dataDir/contre_exemples.json', 'contre_exemples', processContreExemple);
  
  print('\nAll files converted!');
  
  // Print a few samples for verification
  final exFile = File('$dataDir/exercices.json');
  final exData = jsonDecode(exFile.readAsStringSync()) as Map<String, dynamic>;
  final exList = exData['exercices'] as List;
  
  print('\n=== SAMPLE OUTPUTS ===');
  for (int i = 0; i < 3 && i < exList.length; i++) {
    final ex = exList[i];
    print('\n--- ${ex['titre']} ---');
    print('Énoncé: ${ex['enonce']}');
    if (ex['solution'] is List) {
      for (final s in (ex['solution'] as List).take(2)) {
        print('  Sol: $s');
      }
    }
  }
}

File: tool/convert_to_latex.py
#!/usr/bin/env python3
"""Convert Unicode math notation to LaTeX in JSON data files."""

import json
import re
import sys
import os

# Unicode superscript/subscript mappings
SUPERSCRIPTS = {
    '⁰': '0', '¹': '1', '²': '2', '³': '3', '⁴': '4',
    '⁵': '5', '⁶': '6', '⁷': '7', '⁸': '8', '⁹': '9',
    'ⁿ': 'n', 'ᵏ': 'k', 'ᵐ': 'm', 'ᵖ': 'p', 'ᵍ': 'q',
    '⁻': '-', '⁺': '+',
}

SUBSCRIPTS = {
    '₀': '0', '₁': '1', '₂': '2', '₃': '3', '₄': '4',
    '₅': '5', '₆': '6', '₇': '7', '₈': '8', '₉': '9',
    'ₙ': 'n', 'ₖ': 'k', 'ₘ': 'm', 'ₚ': 'p', 'ₐ': 'a',
    'ᵢ': 'i', 'ⱼ': 'j',
}

# Math symbol replacements (inside $...$)
MATH_SYMBOLS = {
    '∈': '\\in ',
    '∉': '\\notin ',
    '≤': '\\leq ',
    '≥': '\\geq ',
    '≠': '\\neq ',
    '≡': '\\equiv ',
    '≅': '\\simeq ',
    '≈': '\\approx ',
    '⟹': '\\Rightarrow ',
    '⟸': '\\Leftarrow ',
    '⟺': '\\Leftrightarrow ',
    '⇒': '\\Rightarrow ',
    '⇐': '\\Leftarrow ',
    '⇔': '\\Leftrightarrow ',
    '→': '\\to ',
    '←': '\\leftarrow ',
    '↦': '\\mapsto ',
    '↗': '\\nearrow ',
    '∀': '\\forall ',
    '∃': '\\exists ',
    '∅': '\\emptyset ',
    '∞': '\\infty',
    '∪': '\\cup ',
    '∩': '\\cap ',
    '⊂': '\\subset ',
    '⊃': '\\supset ',
    '⊆': '\\subseteq ',
    '⊇': '\\supseteq ',
    '⊕': '\\oplus ',
    '⊗': '\\otimes ',
    '⊥': '\\perp ',
    '∎': '\\blacksquare',
    '✓': '',
    '·': '\\cdot ',
    '×': '\\times ',
    'λ': '\\lambda',
    'μ': '\\mu',
    'σ': '\\sigma',
    'π': '\\pi',
    'ε': '\\varepsilon',
    'δ': '\\delta',
    'φ': '\\varphi',
    'ω': '\\omega',
    'θ': '\\theta',
    'α': '\\alpha',
    'β': '\\beta',
    'γ': '\\gamma',
    'Δ': '\\Delta',
    'Σ': '\\Sigma',
    'Φ': '\\Phi',
    'ℓ': '\\ell',
}

# Blackboard bold
BLACKBOARD = {
    'ℝ': '\\mathbb{R}',
    'ℤ': '\\mathbb{Z}',
    'ℚ': '\\mathbb{Q}',
    'ℂ': '\\mathbb{C}',
    'ℕ': '\\mathbb{N}',
    '𝔽': '\\mathbb{F}',
    '𝕋': '\\mathbb{T}',
}

# Indicators and special
SPECIAL = {
    '𝟙': '\\mathbf{1}',
    'ḡ': '\\bar{g}',
    'λ̄': '\\bar{\\lambda}',
}


def collect_superscripts(s, i):
    """Collect consecutive superscript characters starting at position i."""
    result = ''
    while i < len(s) and s[i] in SUPERSCRIPTS:
        result += SUPERSCRIPTS[s[i]]
        i += 1
    return result, i


def collect_subscripts(s, i):
    """Collect consecutive subscript characters starting at position i."""
    result = ''
    while i < len(s) and s[i] in SUBSCRIPTS:
        result += SUBSCRIPTS[s[i]]
        i += 1
    return result, i


def convert_math_expression(text):
    """Convert a string with Unicode math to LaTeX notation."""
    if not text:
        return text
    
    result = []
    i = 0
    in_math = False
    
    while i < len(text):
        ch = text[i]
        
        # Handle superscripts
        if ch in SUPERSCRIPTS:
            sup, new_i = collect_superscripts(text, i)
            if len(sup) > 1:
                result.append('^{' + sup + '}')
            else:
                result.append('^' + sup)
            i = new_i
            continue
        
        # Handle subscripts  
        if ch in SUBSCRIPTS:
            sub, new_i = collect_subscripts(text, i)
            if len(sub) > 1:
                result.append('_{' + sub + '}')
            else:
                result.append('_' + sub)
            i = new_i
            continue
        
        # Handle blackboard bold
        if ch in BLACKBOARD:
            result.append(BLACKBOARD[ch])
            i += 1
            continue
        
        # Handle special symbols
        if ch in SPECIAL:
            result.append(SPECIAL[ch])
            i += 1
            continue
            
        # Handle math symbols
        if ch in MATH_SYMBOLS:
            result.append(MATH_SYMBOLS[ch])
            i += 1
            continue
        
        # Handle ⟨ and ⟩ (angle brackets for inner products)
        if ch == '⟨':
            result.append('\\langle ')
            i += 1
            continue
        if ch == '⟩':
            result.append('\\rangle ')
            i += 1
            continue
        
        # Handle ‖ (norm)
        if ch == '‖':
            result.append('\\|')
            i += 1
            continue
        
        # Handle integral signs with limits
        if ch == '∫':
            result.append('\\int')
            i += 1
            continue
        
        if ch == '∑':
            result.append('\\sum')
            i += 1
            continue
            
        if ch == '∏':
            result.append('\\prod')
            i += 1
            continue
        
        if ch == '√':
            result.append('\\sqrt')
            i += 1
            continue
            
        # Handle floor/ceil
        if ch == '⌊':
            result.append('\\lfloor ')
            i += 1
            continue
        if ch == '⌋':
            result.append('\\rfloor ')
            i += 1
            continue
        
        # Regular character
        result.append(ch)
        i += 1
    
    return ''.join(result)


def wrap_math_regions(text):
    """Wrap mathematical regions of text in $...$ delimiters."""
    if not text:
        return text
    
    # First convert Unicode to LaTeX commands
    text = convert_math_expression(text)
    
    # Now wrap regions that contain LaTeX commands in $...$
    # Strategy: find sequences that contain LaTeX commands and wrap them
    
    # Pattern: sequences containing backslash commands, ^, _, and alphanumeric
    # We need to be smart about what to wrap
    
    # Split by sentence boundaries to process chunks
    # Simple approach: find runs of "math-like" content
    
    # Regex to find math-like regions:
    # - Contains \command
    # - Contains ^ or _
    # - Contains sequences like x^2, A_n, etc.
    
    # First, wrap explicit LaTeX commands that aren't already in $...$
    # Find regions with backslash commands
    
    parts = []
    i = 0
    while i < len(text):
        # Check if we're already inside a math delimiter
        if text[i] == '$':
            # Find matching $
            end = text.find('$', i + 1)
            if end != -1:
                parts.append(text[i:end+1])
                i = end + 1
                continue
        
        # Look for math-like patterns
        # Pattern: word chars possibly followed by subscripts/superscripts and LaTeX commands
        math_match = re.match(
            r'([A-Za-z0-9]*(?:\\[a-zA-Z]+[\{\}A-Za-z0-9]*|[\^_][\{]?[A-Za-z0-9\-\+]+[\}]?)+(?:[A-Za-z0-9\\,\.\+\-\*/\(\)\[\]\{\}=<>!| ]*(?:\\[a-zA-Z]+[\{\}A-Za-z0-9]*|[\^_][\{]?[A-Za-z0-9\-\+]+[\}]?))*)',
            text[i:]
        )
        
        if math_match and len(math_match.group(0).strip()) > 0:
            math_region = math_match.group(0).strip()
            # Only wrap if it contains meaningful LaTeX
            if '\\' in math_region or '^' in math_region or '_' in math_region:
                parts.append('$' + math_region + '$')
                i += math_match.end()
                continue
        
        parts.append(text[i])
        i += 1
    
    result = ''.join(parts)
    
    # Clean up: merge adjacent $...$ regions
    result = re.sub(r'\$\s*\$', ' ', result)
    
    # Clean up: fix $$ at word boundaries  
    result = result.replace('$ $', ' ')
    
    return result


def process_string(text):
    """Process a single string field."""
    if not text or not isinstance(text, str):
        return text
    return wrap_math_regions(text)


def process_list(items):
    """Process a list of strings."""
    if not items or not isinstance(items, list):
        return items
    return [process_string(item) if isinstance(item, str) else item for item in items]


def process_exercise(ex):
    """Process an exercise entry."""
    if 'enonce' in ex:
        ex['enonce'] = process_string(ex['enonce'])
    if 'indication' in ex:
        ex['indication'] = process_string(ex['indication'])
    if 'solution' in ex:
        ex['solution'] = process_list(ex['solution'])
    return ex


def process_demonstration(demo):
    """Process a demonstration entry."""
    if 'titre' in demo:
        demo['titre'] = process_string(demo['titre'])
    if 'enonce' in demo:
        demo['enonce'] = process_string(demo['enonce'])
    if 'preuve' in demo:
        demo['preuve'] = process_list(demo['preuve'])
    if 'corollaires' in demo:
        demo['corollaires'] = process_list(demo['corollaires'])
    if 'remarque' in demo:
        demo['remarque'] = process_string(demo['remarque'])
    return demo


def process_contre_exemple(ce):
    """Process a counter-example entry."""
    if 'enonce' in ce:
        ce['enonce'] = process_string(ce['enonce'])
    if 'explication' in ce:
        ce['explication'] = process_string(ce['explication'])
    return ce


def process_file(filepath, key, processor):
    """Process a JSON file."""
    print(f"Processing {filepath}...")
    
    with open(filepath, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    if key in data:
        data[key] = [processor(item) for item in data[key]]
    
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    
    print(f"  Done! Processed {len(data.get(key, []))} entries.")


if __name__ == '__main__':
    base = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    data_dir = os.path.join(base, 'assets', 'data')
    
    process_file(
        os.path.join(data_dir, 'exercices.json'),
        'exercices', process_exercise
    )
    
    process_file(
        os.path.join(data_dir, 'demonstrations.json'),
        'demonstrations', process_demonstration
    )
    
    process_file(
        os.path.join(data_dir, 'contre_exemples.json'),
        'contre_exemples', process_contre_exemple
    )
    
    print("\nAll files converted!")
